### PR TITLE
SDS agent bring clusterID meta when send csr request

### PR DESCRIPTION
**Pull request by hzxuzhonghu**
This is for single controlplane multi-cluster authentication.

### CODE REVIEW

**Comment by hzxuzhonghu** [comment]
cc @myidpt @irisdingbj

**Comment by irisdingbj** [comment]
should we add clusterID into configuration file? Now this is only valid under multi-cluster:
```
  multiCluster:
    # Set to true to connect two kubernetes clusters via their respective
    # ingressgateway services when pods in each cluster cannot directly
    # talk to one another. All clusters should be using Istio mTLS and must
    # have a shared root CA for this model to work.
    enabled: false
    # Should be set to the name of the cluster this installation will run in. This is required for sidecar injection
    # to properly label proxies
    clusterName: ""
```

**Comment by lei-tang** [comment]
@hzxuzhonghu Can you add a description of the workflows involved with this PR to the PR description or to the code? 
1) Where does the clusterID meta originate from (e.g., is it from a configuration? How to configure?)? 
2) Where is the clusterID information consumed? Will the clusterID information become a field in the certificate signed?

**Comment by hzxuzhonghu** [comment]
@lei-tang This is not a complete pr. 

What i do is the first step:
The clusterID is from an env `ISTIO_META_CLUSTER_ID` which is injected to sidecar. 

```
    - name: ISTIO_META_CLUSTER_ID
      value: "{{ valueOrDefault .Values.global.multiCluster.clusterName `Kubernetes` }}"
```


I will supplement the second step, parse the clusterid and send token review only to the ssspecific cluster rather than try all clusters.

**Comment by hzxuzhonghu** [comment]
In istiod we can parse it as what we do for the `token`. To be exact, it is not used in CSR request, it is for csr request auth.

File: pilot/cmd/pilot-agent/main.go [hunk 1]
```diff
@@ -90,6 +90,7 @@ var (
 	podNamespaceVar      = env.RegisterStringVar("POD_NAMESPACE", "", "")
 	istioNamespaceVar    = env.RegisterStringVar("ISTIO_NAMESPACE", "", "")
 	kubeAppProberNameVar = env.RegisterStringVar(status.KubeAppProberEnvName, "", "")
+	clusterIDVar         = env.RegisterStringVar("ISTIO_META_CLUSTER_ID", "", "")
```
Comment (costinm) [inline pilot/cmd/pilot-agent/main.go:93]: Don't we have it in proxy config already ? 

I think @howardjohn moved almost everything to use ProxyConfig - with proper annotation to override, etc.

At this point we shouldn't add more env variables - at least not for important things like cluster ID.

File: security/pkg/server/ca/authenticate/authenticator.go [hunk 2]
```diff
@@ -28,9 +28,10 @@ import (
 )
 
 const (
-	bearerTokenPrefix = "Bearer "
-	httpAuthHeader    = "authorization"
-	idTokenIssuer     = "https://accounts.google.com"
+	bearerTokenPrefix   = "Bearer "
+	httpAuthHeader      = "authorization"
+	httpClusterIDHeader = "clusterid"
```
Comment (costinm) [inline security/pkg/server/ca/authenticate/authenticator.go:33]: Do we want to pass it in a HTTP header ? I guess it's a quick hack, but most APIs put this in request.

**Comment by costinm** [comment]
Thanks for doing this work. I left few cosmetic comments - but we can iterate in separate PRs, nothing blocking IMO.

File: pilot/cmd/pilot-agent/main.go (hunk 1 above)
Comment (costinm) [reply pilot/cmd/pilot-agent/main.go:93]: I'm almost certain we have some other ways to pass the cluster ID ( or registry ID actually ), please check with @howardjohn - he fixed some related bug. Sorry, my memory is bad.

Comment (hzxuzhonghu) [reply pilot/cmd/pilot-agent/main.go:93]: This is not a new env.

File: security/pkg/server/ca/authenticate/authenticator.go (hunk 2 above)
Comment (hzxuzhonghu) [reply security/pkg/server/ca/authenticate/authenticator.go]: no，fixed

**Comment by hzxuzhonghu** [comment]
ping @myidpt to have a look represent @istio/wg-security-maintainers

**Comment by myidpt** [comment]
General question: why can't we use the issuer in the K8s JWT to distinguish the different clusters?

**Comment by costinm** [comment]
K8S is not required to include a cluster name or different issuer. GCP
might - but this is not only for GCP.

File: security/pkg/nodeagent/caclient/providers/citadel/client_test.go [hunk 3]
```diff
@@ -88,7 +88,7 @@ func TestCitadelClient(t *testing.T) {
 		// The goroutine starting the server may not be ready, results in flakiness.
 		time.Sleep(1 * time.Second)
 
-		cli, err := NewCitadelClient(lis.Addr().String(), false, nil)
+		cli, err := NewCitadelClient(lis.Addr().String(), false, nil, "")
```
[the remark this answers is not in the document]
Comment (hzxuzhonghu) [reply security/pkg/nodeagent/caclient/providers/citadel/client_test.go:91]: We donot have a test for single control plane multi-cluster yet

File: security/pkg/server/ca/authenticate/kube_jwt_test.go [hunk 4]
```diff
@@ -15,106 +15,43 @@
...
 		metadata       metadata.MD
-		client         tokenReviewClient
+		jwtPolicy      string
 		expectedID     string
 		expectedErrMsg string
 	}{
 		"No bearer token": {
 			metadata: metadata.MD{
-				"random": []string{},
+				"clusterid": []string{"Kubernetes"},
```
[the remark this answers is not in the document]
Comment (hzxuzhonghu) [reply security/pkg/server/ca/authenticate/kube_jwt_test.go:54]: ok

File: pilot/pkg/serviceregistry/kube/controller/multicluster.go [hunk 5]
```diff
@@ -168,14 +168,11 @@ func (m *Multicluster) updateHandler(svc *model.Service) {
...
-func (m *Multicluster) GetRemoteKubeClients() map[string]kubernetes.Interface {
+func (m *Multicluster) GetRemoteKubeClient(clusterID string) kubernetes.Interface {
 	m.m.Lock()
-	res := make(map[string]kubernetes.Interface)
-	for k, v := range m.remoteKubeControllers {
-		if v != nil && v.rc != nil {
-			res[k] = v.rc.client
-		}
+	if c := m.remoteKubeControllers[clusterID]; c != nil {
+		return c.client
```
[the remark this answers is not in the document]
Comment (hzxuzhonghu) [reply pilot/pkg/serviceregistry/kube/controller/multicluster.go:174]: good catch

File: security/pkg/nodeagent/caclient/providers/citadel/client_test.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (hzxuzhonghu) [reply security/pkg/nodeagent/caclient/providers/citadel/client_test.go:91]: IC, a ut is neccessary

**Comment by hzxuzhonghu** [comment]
done, added test cases

**Comment by hzxuzhonghu** [comment]
@myidpt any other comments

File: pilot/cmd/pilot-agent/main.go (hunk 1 above)
Comment (hovonako) [reply pilot/cmd/pilot-agent/main.go:93]: Hi @hzxuzhonghu and @costinm 
I found a ClusterID at "pilot/pkg/model/context.go"(type Proxy struct) with comment:
`
type Proxy struct {
	// ClusterID specifies the cluster where the proxy resides.
	// TODO: clarify if this is needed in the new 'network' model, likely needs to
	// be renamed to 'network'
	ClusterID string
`
It's seems that model contains all meta related to the proxy network (include ClusterID)

Comment (hzxuzhonghu) [reply pilot/cmd/pilot-agent/main.go:93]: @hovonako That one in the NodeMeta indicates the cluster for proxy, here it is where the istiod resides.

**Pull request merged**